.pre-commit-config.yaml

```diff
@@ -0,0 +1,7 @@
+repos:
+-   repo: https://github.com/pre-commit/mirrors-clang-format
+    rev: 'v14.0.0'  # The default in Ubuntu 22.04, which is used in our CI
```

Comment:
I don't really know, how the user environment is changes. But what I can tell you is that on my system, where I use pre-commit, I don't have clang-14 in my path, even though it is used by pre-commit.

Reply:
After some search, it is handled by python virtualenv.
The default folder is ~/.cache/pre-commit `https://[pre-commit](https://pre-commit.com/#managing-ci-caches).com/#managing-ci-caches`
there's a folder (py_env-python3.9) under a random folder, `source py_env-python3.9/bin/activate` can run clang-format from the environment.
I am still looking for how to get the path correctly and use it in editor

Reply:
For what do you need it in your editor? You can just run `pre-commit run` and it will do the formatting.

Reply:
I usually run several times clang-format during coding with the default shortcut.
pre-commit requires us to run via command and pass files (or all-files)

Reply:
It may work in almost case but failed with the comment and string because clang-format does not try to concat them back if break early.
local clang-format breaks the string/comment eariler due to different indentation or format. the pre-commit does not fix it if they are still under 80 column limit (but not correct)

Reply:
I will check that out.

Reply:
So, I don't think there is a clang-format option for that. There is `ReflowComments`, but that only splits up long lines, and does not concat short lines, see: https://clang.llvm.org/docs/ClangFormatStyleOptions.html#reflowcomments. All other clang-format options related to comments seem not to do the concatenation. Which clang-format version are you using?

Reply:
No, I mean no-concatenation will introduce issue from inconsistent clang-format.
local clang-format may break some comments which are considered long with this format. but pre-commit one does not break in this way.
For example, assuming `pre-commit` clang-format only put the normal indentation for arguments, but local clang-format aligns the arguments
1 is the original code
```
void func(...,
          tttt // some text ...
```
2 is from local clang-format (break when it is too long)
```
void func(...,
          tttt // some text
               // more text
```
3 is format 1 from pre-commit clang-format, less indention so no need for breaking
```
void func(...,
    tttt // some text ...
```
4 is format 2 from pre-commit clang-format, no-concatenation if it is already split
```
void func(...,
    tttt // some text
         // more text
```
3 and 4 are different. If we apply format directly, we should get 3. 
With inconsistent clang-format, we may get 4 in the end.

Reply:
can you still provide your clang-format version? I will still look into the issue you mentioned.

Reply:
TBH, I think this might be one of those issues which just change depending on the clang format versions.
Just as an example, I used this:
```
    LinOp* apply(ptr_param<const LinOp> alpha,
                 ptr_param<const LinOp> b, // tis a long comment describing the parametr b for what ever reasons
                 ptr_param<const LinOp> beta,
                 ptr_param<LinOp> x)
```
and the pre-commit gave me:
```
    LinOp* apply(ptr_param<const LinOp> alpha,
                 ptr_param<const LinOp> b,  // tis a long comment describing the
                                            // parametr b for what ever reasons
                 ptr_param<const LinOp> beta, ptr_param<LinOp> x)
```
which is your 2. But maybe it depends more on the actual example.

Reply:
The issue happens when the different indention format clang-format, that's why the example is under the assumption.
In more realistic example, you need to look for the attribute usage.
Clang-format does not give any consistence between version if no option to control behavior such that we may have different indention. It's also why we need to fix a version for repo code not just rely on the config version requirement. There are always some undefined behavior across versions.
Otherwise, the config only requires 9+ and format_header only requires 12+. 
clang-format 9, 12, 14, 17

Reply:
I'm not sure if I understand your point. The whole point of this PR is to fix a clang-format version that is used both in the CI and each local repo.